README.rst
SoundMeter
==========

.. image:: https://img.shields.io/pypi/v/soundmeter.png
        :target: https://pypi.python.org/pypi/soundmeter

SoundMeter is a command-line tool to obtain sound power in real time. It basically turns the audio recording functionality into a sound meter for machines that ship with audio input devices (e.g. microphone). It currently reveals the `root-mean-square (RMS) of sound fragments <http://docs.python.org/2/library/audioop.html#audioop.rms>`_, which is a measure of the power in an audio signal. The actual values also depend on the system settings of sound input.

.. image:: https://asciinema.org/a/RdNCvGdsdvxdu8O9nLqkhhgaT.png
        :target: https://asciinema.org/a/RdNCvGdsdvxdu8O9nLqkhhgaT

Features
--------

- A command-line meter that supports triggering upon events
- Monitor API for backend module programming

Dependencies
------------
System packages
~~~~~~~~~~~~~~~
On Debian/Ubuntu::

    $ sudo apt-get install portaudio19-dev python-dev alsa-utils

On Fedora/RHEL::

    $ sudo yum install portaudio-devel python-devel alsa-utils

On OS X::

    $ brew install portaudio

PyPI packages
~~~~~~~~~~~~~
- argparse
- pyaudio
- pydub
- python-daemon


Installation
------------

You can install the package with pip using the following command to allow externally hosted packages::

  $ pip install soundmeter --allow-all-external --allow-unverified pyaudio

Or, you can download a source distribution and install with these commands::

  $ python setup.py install


Usage
-----
The simplest usage is to run "soundmeter" from command-line without any options::

    $ soundmeter

Collect RMS values for 10 seconds so that you will know the sound level in the current environment::

    $ soundmeter --collect --seconds 10
    Collecting RMS values...
       154  Timeout
    Collected result:
        min:        152
        max:        211
        avg:        156

You can set trigger and action for soundmeter.

Stop the soundmeter if RMS is greater than 211 consecutively for 3 times::

    $ soundmeter --trigger +211 3 --action stop

Execute ``trigger.sh`` if RMS is greater than 211::

    $ soundmeter --trigger +211 --action exec --exec trigger.sh

Execute ``trigger.sh`` and stop soundmeter if RMS is less than 152 consecutively for 3 times::

    $ soundmeter --trigger -152 3 --action exec-stop --exec trigger.sh

Run the soundmeter with trigger and action in the background::

    $ soundmeter --trigger +211 3 --action exec --exec trigger.sh --daemonize

Run the soundmeter for 2 minutes and log to ``meter.log``::

    $ soundmeter --seconds 120 --log meter.log

Command-line Options
~~~~~~~~~~~~~~~~~~~~

The "soundmeter" command accepts the following options:

  -p PROFILE, --profile PROFILE  config profile (section name)
  -c, --collect  collect RMS values to determine thresholds
  -s SECS, --seconds SECS  time in seconds to run the meter (default forever)
  -a ACTION_TYPE, --action ACTION_TYPE  triggered action (``stop``, ``exec-stop`` and ``exec``)
  -t THRESHOLD, --trigger THRESHOLD  trigger condition (threshold RMS and an optional number of consecutive triggering times, which defaults 1)
  -e FILE, --execute FILE  shell script to execute upon trigger (defaults to ``~/.soundmeter/trigger.sh``), can be configured to pass the "last triggering" RMS value as argument by setting ``rms_as_trigger_arg`` to ``True`` in ~/.soundmeter/config
  -d, --daemonize  run the meter in the background
  --log LOGFILE  log the meter (defaults to ``~/.soundmeter/log``)
  -v, --verbose         verbose mode
  --segment SECONDS  audio segment length recorded in seconds (defaults to 0.5); when specified, it overrides ``audio_segment_length`` in ~/.soundmeter/config

Config
~~~~~~
Some "dependency-required" parameters can be configured at ~/.soundmeter/config. The default configuration is::

    [soundmeter]
    frames_per_buffer = 2048
    format = 8
    channels = 2
    rate = 44100
    audio_segment_length = 0.5
    rms_as_trigger_arg = False

You can have multiple sections in the config file and specify the one to use with the ``--profile`` command-line options. The default profile name is ``soundmeter``. For example::

    [soundmeter]
    frames_per_buffer = 2048
    format = 8
    channels = 2

    [test]
    frames_per_buffer = 1024
    format = 8
    channels = 1

To use the ``test`` profile::

    $ soundmeter --profile test ...

There is also an ``input_device_index`` parameter, which specifies the index of input device to use. If unspecified, it uses the default input device.


CHANGELOG
v0.1.5
======

- Support multiple config sections and add -p/--profile command-line option.

v0.1.4
======

- Add support to specify input device index.

v0.1.3
======

- Supported RMS value to be passed as an argument to trigger scripts.
- Supported Python 3.
- Removed `daemonize` arguments from Meter class.

v0.1
====

- Initial release.


soundmeter/tests.py
import os
import signal
import subprocess
import time
from unittest import TestCase
from scripttest import TestFileEnvironment
from .cli import setup_user_dir
from .meter import Meter
from .utils import create_executable


d = os.path.dirname(__file__)
PROJECT_PATH = os.path.abspath(os.path.join(d, os.pardir))


def create_run_script():
    run_script = os.path.join(PROJECT_PATH, 'run.py')
    content = '#!/usr/bin/env python\n'
    content += 'from soundmeter.meter import main\n\n\n'
    content += 'main()'
    if not os.path.exists(run_script):
        create_executable(run_script, content)


setup_user_dir()
create_run_script()


class TestMeter(TestCase):
    """Test Meter class programmatically"""
    def setUp(self):
        self.meter = Meter(seconds=2.0)

    def test_running(self):
        self.assertFalse(self.meter.is_running)
        self.meter.start()
        self.assertFalse(self.meter.is_running)


class TestBasicCommands(TestCase):
    """Test basic command-line invoke of the program"""
    def setUp(self):
        self.env = TestFileEnvironment('./test-output')

    def test_default(self):
        res = self.env.run('../run.py', '-s', '1',
                           expect_stderr=True)
        assert 'Timeout' in res.stdout
        self.assertEqual(res.returncode, 0)

    def test_collect(self):
        res = self.env.run('../run.py', '-s', '1', '-c',
                           expect_stderr=True)
        assert 'Collecting' in res.stdout
        self.assertEqual(res.returncode, 0)

    def test_log(self):
        res = self.env.run('../run.py', '-s', '1', '--log', 'log.txt',
                           expect_stderr=True)
        assert 'Timeout' in res.stdout
        assert 'log.txt' in res.files_created
        self.assertEqual(res.returncode, 0)

    def test_segment(self):
        res = self.env.run('../run.py', '-s', '1', '--segment', '0.2',
                           expect_stderr=True)
        assert 'Timeout' in res.stdout
        self.assertEqual(res.returncode, 0)

    def tearDown(self):
        pass


class TestCommands(TestCase):
    def test_sigint(self):
        popen = subprocess.Popen(['./run.py'])
        time.sleep(2)
        os.kill(popen.pid, signal.SIGINT)

    def test_arguments(self):
        popen = subprocess.Popen(['./run.py', '-t', '10000', '-a', 'stop'])
        time.sleep(2)
        os.kill(popen.pid, signal.SIGINT)

    def test_daemon(self):
        popen = subprocess.Popen(['./run.py', '-d'], shell=True)
        time.sleep(2)
        popen.send_signal(signal.SIGINT)


class TestConfig(TestCase):
    def test_config(self):
        config = os.path.join(PROJECT_PATH, 'sample_config')
        os.environ['SOUNDMETER_TEST_CONFIG'] = config
        popen = subprocess.Popen(['./run.py'], env=os.environ.copy())
        time.sleep(2)
        popen.send_signal(signal.SIGINT)


soundmeter/meter.py
from __future__ import print_function
import daemon
import logging
import pyaudio
import pydub
import wave
import signal
import six
import subprocess
import sys
import time
import warnings
if six.PY2:
    try:
        from cStringIO import StringIO
    except ImportError:
        from StringIO import StringIO
else:
    from io import BytesIO as StringIO
from .settings import Config
from .cli import get_meter_kwargs, setup_user_dir
from .utils import noalsaerr, coroutine


__all__ = ['Meter']
warnings.filterwarnings("ignore", category=DeprecationWarning)
_soundmeter = None


class Meter(object):

    class StopException(Exception):
        pass

    def __init__(self, collect=False, seconds=None, action=None,
                 threshold=None, num=None, script=None, log=None,
                 verbose=False, segment=None, profile=None, *args, **kwargs):
        """
        :param bool collect: A boolean indicating whether collecting RMS values
        :param float seconds: A float representing number of seconds to run the
            meter (None for forever)
        :param str action: The action type ('stop', 'exec-stop' or 'exec')
        :param str threshold: A string representing threshold and bound type
            (e.g. '+252', '-144')
        :param int num: An integer indicating how many consecutive times the
            threshold is reached before triggering the action
        :param script: File object representing the script to be executed
        :param log: File object representing the log file
        :param bool verbose: A boolean for verbose mode
        :param float segment: A float representing `AUDIO_SEGMENT_LENGTH`
        :param str profile: The config profile
        """

        global _soundmeter
        _soundmeter = self  # Register this meter globally
        self.config = Config(profile)
        self.output = StringIO()
        with noalsaerr():
            self.audio = pyaudio.PyAudio()
        self.stream = self.audio.open(
            format=self.config.FORMAT,
            channels=self.config.CHANNELS,
            input_device_index=self.config.INPUT_DEVICE_INDEX,
            input=True,
            rate=self.config.RATE,
            frames_per_buffer=self.config.FRAMES_PER_BUFFER)
        self.collect = collect
        self.seconds = seconds
        self.action = action
        self.threshold = threshold
        self.num = num
        self.script = script
        self.log = log
        self.verbose = verbose
        self.segment = segment
        self.is_running = False
        self._graceful = False  # Graceful stop switch
        self._timeout = False
        self._timer = None
        self._data = {}
        self._setup_logging()

    @coroutine
    def record(self):
        """
        Record PyAudio stream into StringIO output

        This coroutine keeps stream open; the stream is closed in stop()
        """

        while True:
            frames = []
            self.stream.start_stream()
            for i in range(self.num_frames):
                data = self.stream.read(self.config.FRAMES_PER_BUFFER)
                frames.append(data)
            self.output.seek(0)
            w = wave.open(self.output, 'wb')
            w.setnchannels(self.config.CHANNELS)
            w.setsampwidth(self.audio.get_sample_size(self.config.FORMAT))
            w.setframerate(self.config.RATE)
            w.writeframes(b''.join(frames))
            w.close()
            yield

    def start(self):
        segment = self.segment or self.config.AUDIO_SEGMENT_LENGTH
        self.num_frames = int(
            self.config.RATE / self.config.FRAMES_PER_BUFFER * segment)
        if self.seconds:
            signal.setitimer(signal.ITIMER_REAL, self.seconds)
        if self.verbose:
            self._timer = time.time()
        if self.collect:
            print('Collecting RMS values...')
        if self.action:
            # Interpret threshold
            self.get_threshold()

        try:
            self.is_running = True
            record = self.record()
            while not self._graceful:
                record.send(True)  # Record stream `AUDIO_SEGMENT_LENGTH' long
                data = self.output.getvalue()
                segment = pydub.AudioSegment(data)
                rms = segment.rms
                if self.collect:
                    self.collect_rms(rms)
                self.meter(rms)
                if self.action:
                    if self.is_triggered(rms):
                        self.execute(rms)
                self.monitor(rms)
            self.is_running = False
            self.stop()

        except self.__class__.StopException:
            self.is_running = False
            self.stop()

    def meter(self, rms):
        if not self._graceful:
            sys.stdout.write('\r%10d  ' % rms)
            sys.stdout.flush()
            if self.log:
                self.logging.info(rms)

    def graceful(self):
        """Graceful stop so that the while loop in start() will stop after the
         current recording cycle"""
        self._graceful = True

    def timeout(self):
        msg = 'Timeout'
        print(msg)
        if self.log:
            self.logging.info(msg)
        self.graceful()

    def stop(self):
        """Stop the stream and terminate PyAudio"""
        self.prestop()
        if not self._graceful:
            self._graceful = True
        self.stream.stop_stream()
        self.audio.terminate()
        msg = 'Stopped'
        self.verbose_info(msg, log=False)
        # Log 'Stopped' anyway
        if self.log:
            self.logging.info(msg)
        if self.collect:
            if self._data:
                print('Collected result:')
                print('    min: %10d' % self._data['min'])
                print('    max: %10d' % self._data['max'])
                print('    avg: %10d' % int(self._data['avg']))
        self.poststop()

    def get_threshold(self):
        """Get and validate raw RMS value from threshold"""

        if self.threshold.startswith('+'):
            if self.threshold[1:].isdigit():
                self._threshold = int(self.threshold[1:])
                self._upper = True
        elif self.threshold.startswith('-'):
            if self.threshold[1:].isdigit():
                self._threshold = int(self.threshold[1:])
                self._upper = False
        else:
            if self.threshold.isdigit():
                self._threshold = int(self.threshold)
                self._upper = True
        if not hasattr(self, '_threshold'):
            raise ValueError('Invalid threshold')

    def is_triggered(self, rms):
        if self._upper and rms > self._threshold \
                or not self._upper and rms < self._threshold:
            if 'triggered' in self._data:
                self._data['triggered'] += 1
            else:
                self._data['triggered'] = 1
        else:
            if 'triggered' in self._data:
                del self._data['triggered']
        t = self._data.get('triggered')
        if isinstance(t, int) and t >= self.num:
            return True
        return False

    def execute(self, rms):
        if self.action == 'stop':
            msg = 'Stop Action triggered'
            print(msg)
            if self.log:
                self.logging.info(msg)
            raise self.__class__.StopException('stop')

        elif self.action == 'exec-stop':
            msg = 'Exec-Stop Action triggered'
            print(msg)
            if self.log:
                self.logging.info(msg)
            v = 'Executing %s' % self.script
            self.verbose_info(v)
            self.popen(rms)
            raise self.__class__.StopException('exec-stop')

        elif self.action == 'exec':
            msg = 'Exec Action triggered'
            print(msg)
            if self.log:
                self.logging.info(msg)
            v = 'Executing %s' % self.script
            self.verbose_info(v)
            self.popen(rms)

    def popen(self, rms):
        self.prepopen()
        if self.script:
            try:
                cmd = [self.script]
                """If configured as True, rms value is passed
                as an argument for the script"""
                if (self.config.RMS_AS_TRIGGER_ARG):
                    cmd.append(str(rms))
                subprocess.Popen(cmd)
            except OSError as e:
                msg = 'Cannot execute the shell script: %s' % e
                print(msg)
                if self.log:
                    self.logging.info(msg)
        self.postpopen()

    def collect_rms(self, rms):
        """Collect and calculate min, max and average RMS values"""
        if self._data:
            self._data['min'] = min(rms, self._data['min'])
            self._data['max'] = max(rms, self._data['max'])
            self._data['avg'] = float(rms + self._data['avg']) / 2
        else:
            self._data['min'] = rms
            self._data['max'] = rms
            self._data['avg'] = rms

    def verbose_info(self, msg, log=True):
        if self.verbose:
            print(msg)
            if self.log and log:
                self.logging.info(msg)

    def _setup_logging(self):
        if self.log:
            self.logging = logging.basicConfig(
                filename=self.log, format='%(asctime)s %(message)s',
                level=logging.INFO)
            self.logging = logging.getLogger(__name__)

    def monitor(self, rms):
        """This function is to be overridden"""
        pass

    def prepopen(self):
        """Pre-popen hook"""
        pass

    def postpopen(self):
        """Post-popen hook"""
        pass

    def prestop(self):
        """Pre-stop hook"""
        pass

    def poststop(self):
        """Post-stop hook"""
        pass

    def __repr__(self):
        u = self.action if self.action else 'no-action'
        return '<%s: %s>' % (self.__class__.__name__, u)


def main():
    setup_user_dir()
    kwargs = get_meter_kwargs()
    if kwargs.pop('daemonize'):
        daemon_context = daemon.DaemonContext()
        # python-daemon>=2.1 has initgroups=True by default but it requires
        # root privileges.
        # setting daemon_context.initgroups to False instead of passing
        # arguments to daemon.DaemonContext will not break older versions
        daemon_context.initgroups = False
        with daemon_context:
            m = Meter(**kwargs)
            m.start()
    else:
        m = Meter(**kwargs)
        m.start()


# Signal handlers
def sigint_handler(signum, frame):
    sys.stdout.write('\n')
    _soundmeter.graceful()


def sigalrm_handler(signum, frame):
    _soundmeter.timeout()


# Register signal handlers
signal.signal(signal.SIGINT, sigint_handler)
signal.signal(signal.SIGALRM, sigalrm_handler)


soundmeter/utils.py
from ctypes import *  # NOQA
from contextlib import contextmanager
import os
import six
import stat


def get_file_path(f):
    if f:
        name = getattr(f, 'name')
        if name:
            path = os.path.abspath(name)
            return path


def create_executable(path, content):
    with open(path, 'w') as f:
        f.write(content)
    s = os.stat(path)
    os.chmod(path, s.st_mode | stat.S_IEXEC)


# Work-around on error messages by alsa-lib
# http://stackoverflow.com/questions/7088672/
ERROR_HANDLER_FUNC = CFUNCTYPE(None, c_char_p, c_int,
                               c_char_p, c_int, c_char_p)


def py_error_handler(filename, line, function, err, fmt):
    pass


c_error_handler = ERROR_HANDLER_FUNC(py_error_handler)


@contextmanager
def noalsaerr():
    try:
        asound = cdll.LoadLibrary('libasound.so')
        asound.snd_lib_error_set_handler(c_error_handler)
        yield
        asound.snd_lib_error_set_handler(None)
    except:
        yield


def coroutine(func):
    def start(*args, **kwargs):
        g = func(*args, **kwargs)
        if six.PY2:
            g.next()
        else:
            g.__next__()
        return g
    return start


soundmeter/command.py
import argparse
import os
import stat
import sys
from .settings import PROG, USER_LOGFILE, USER_SCRIPT, USER_DIR
from .utils import get_file_path


def parse_args():
    parser = argparse.ArgumentParser(usage='%(prog)s [options]',
                                     prog=PROG)
    parser.add_argument('-c', '--collect', action='store_true',
                        help='collect RMS values to determine thresholds')
    seconds_help = 'time in seconds to run the meter (default forever)'
    parser.add_argument('-s', '--seconds', type=float,
                        help=seconds_help)
    parser.add_argument('-a', '--action',
                        choices=['stop', 'exec-stop', 'exec'],
                        help="triggered action")
    trigger_help = 'trigger condition (threshold RMS and number of times)'
    parser.add_argument('-t', '--trigger', nargs='+',
                        metavar=('[+|-]THRESHOLD', 'NUM'),
                        help=trigger_help)
    parser.add_argument('-e', '--exec', dest='script',
                        metavar='SCRIPT', type=argparse.FileType('r'),
                        default=USER_SCRIPT,
                        help='shell script to execute upon trigger')
    parser.add_argument('-d', '--daemonize', action='store_true',
                        help='run the meter in the background')
    parser.add_argument('--log', nargs='?', metavar='LOGFILE',
                        type=argparse.FileType('a'),
                        const=USER_LOGFILE,
                        help='log the meter (default to ~/.soundmeter/log)')
    parser.add_argument('-v', '--verbose', action='store_true',
                        help='verbose mode')
    segment_help = 'audio segment length recorded in seconds (defaults to 0.5)'
    parser.add_argument('--segment', metavar='SECONDS', help=segment_help)

    # Extra validation of arguments
    args = parser.parse_args()
    if args.collect:
        if args.action or args.trigger:
            msg = ('-c/--collect should not be used with -a/--action '
                   'or -t/--trigger')
            raise parser.error(msg)
    if args.segment:
        try:
            segment = float(args.segment)
        except ValueError:
            msg = '--segment must be an integer or float'
            raise parser.error(msg)
        if segment < 0.05:
            msg = '--segment cannot be smaller than 0.05'
            raise parser.error(msg)
    if args.action:
        if not args.trigger:
            msg = 'must specify -t/--trigger when using -a/--action'
            raise parser.error(msg)
        if args.action == 'stop':
            args.script = None
        if len(args.trigger) == 1:
            args.trigger.append('1')
        elif len(args.trigger) > 2:
            trigger_msg = '-t/--trigger accepts at most two arguments'
            raise parser.error(trigger_msg)
        else:
            trigger_msg = ('the second argument NUM to -t/--trigger must be an'
                           ' positive integer')
            if not args.trigger[1].isdigit():
                raise parser.error(trigger_msg)
            if args.trigger[1].isdigit() and int(args.trigger[1]) == 0:
                raise parser.error(trigger_msg)

    elif args.trigger:
        msg = 'must specify -a/--action when using -t/--trigger'
        raise parser.error(msg)

    elif args.script:
        if '-e' in sys.argv or '--exec' in sys.argv:
            msg = ('must specify -a/--action and -t/--trigger when using '
                   '-e/--exec')
            raise parser.error(msg)
        else:
            args.script = None
    return args


def get_meter_kwargs():
    kwargs = dict(parse_args()._get_kwargs())
    # Convert `trigger' into `threshold' and `num'
    if kwargs['trigger'] is not None:
        kwargs['threshold'] = kwargs['trigger'][0]
        kwargs['num'] = int(kwargs['trigger'][1])
    del kwargs['trigger']
    kwargs['script'] = get_file_path(kwargs['script'])
    kwargs['log'] = get_file_path(kwargs['log'])
    kwargs['segment'] = float(kwargs['segment']) \
        if kwargs['segment'] is not None else None
    return kwargs


def setup_user_dir():
    if not os.path.exists(USER_DIR):
        os.makedirs(USER_DIR)
    if not os.path.exists(USER_SCRIPT):
        content = '!/bin/sh\n'
        create_executable(USER_SCRIPT, content)


def create_executable(path, content):
    with open(path, 'w') as f:
        f.write(content)
    s = os.stat(path)
    os.chmod(path, s.st_mode | stat.S_IEXEC)


soundmeter/monitor.py
from .meter import Meter


class Monitor(Meter):

    def __init__(self, *args, **kwargs):
        super(Monitor, self).__init__(*args, **kwargs)

    def monitor(self, rms):
        """Extra monitor actions with RMS values

        Custom trigger conditions and actions can be written here as long as
        `action=None` is passed to the Monitor(). The Monitor instance will
        only execute monitor() as action is disabled.
        """
        pass

    # The following hooks can be used with as long as
    # `action=None` is passed to the Monitor()
    def prepopen(self):
        """Extra code before executing the script"""
        pass

    def postpopen(self):
        """Extra code after executing the script"""
        pass

    # The following hooks can be used in all cases
    def prestop(self):
        """Extra code before stop"""
        pass

    def poststop(self):
        """Extra code after stop"""
        pass


soundmeter/settings.py
try:
    import configparser
except ImportError:
    import ConfigParser as configparser
import os
import pyaudio


PROG = 'soundmeter'
USER_DIR = os.path.join(os.path.expanduser('~'), '.' + PROG)
USER_LOGFILE = os.path.join(USER_DIR, 'log')
USER_CONFIG = os.path.join(USER_DIR, 'config')
USER_SCRIPT = os.path.join(USER_DIR, 'trigger.sh')


class Config(object):

    FRAMES_PER_BUFFER = 2048
    FORMAT = pyaudio.paInt16
    CHANNELS = 2
    INPUT_DEVICE_INDEX = None
    RATE = 44100
    AUDIO_SEGMENT_LENGTH = 0.5
    RMS_AS_TRIGGER_ARG = False

    def __init__(self, section=None):
        config = configparser.ConfigParser()
        config.read(os.environ.get('SOUNDMETER_TEST_CONFIG') or USER_CONFIG)
        items = {}

        if section is None:
            if config.has_section(PROG):
                section = PROG
            else:
                return

        if config.has_section(section):
            items = dict(config.items(section))
            for name in items:
                try:
                    if name in ['frames_per_buffer', 'format', 'channels',
                                'rate', 'input_device_index']:
                        items[name] = int(items[name])
                    elif name in ['audio_segment_length']:
                        items[name] = float(items[name])
                    elif name in ['rms_as_trigger_arg']:
                        items[name] = bool(items[name])
                    else:
                        msg = \
                            'Unknown name "%s" in config section "%s"' % (
                                name, section)
                        raise Exception(msg)
                except ValueError:
                    msg = \
                        'Invalid value to "%s" in config section "%s"' % (
                                name, section)
                    raise Exception(msg)
        else:
            raise Exception('No section named "%s" in config' % section)

        for name, value in items.items():
            setattr(self, name.upper(), value)


soundmeter/cli.py
from __future__ import print_function
import argparse
import os
import sys
from .settings import PROG, USER_LOGFILE, USER_SCRIPT, USER_DIR
from .utils import get_file_path, create_executable


def parse_args():
    parser = argparse.ArgumentParser(usage='%(prog)s [options]',
                                     prog=PROG)
    parser.add_argument('-p', '--profile',
                        help='config profile (section name)')
    parser.add_argument('-c', '--collect', action='store_true',
                        help='collect RMS values to determine thresholds')
    seconds_help = 'time in seconds to run the meter (default forever)'
    parser.add_argument('-s', '--seconds', type=float,
                        help=seconds_help)
    parser.add_argument('-a', '--action',
                        choices=['stop', 'exec-stop', 'exec'],
                        help="triggered action")
    trigger_help = 'trigger condition (threshold RMS and number of times)'
    parser.add_argument('-t', '--trigger', nargs='+',
                        metavar=('[+|-]THRESHOLD', 'NUM'),
                        help=trigger_help)
    parser.add_argument('-e', '--exec', dest='script',
                        metavar='SCRIPT', type=argparse.FileType('r'),
                        default=USER_SCRIPT,
                        help='shell script to execute upon trigger')
    parser.add_argument('-d', '--daemonize', action='store_true',
                        help='run the meter in the background')
    parser.add_argument('--log', nargs='?', metavar='LOGFILE',
                        type=argparse.FileType('a'),
                        const=USER_LOGFILE,
                        help='log the meter (default to ~/.soundmeter/log)')
    parser.add_argument('-v', '--verbose', action='store_true',
                        help='verbose mode')
    segment_help = 'audio segment length recorded in seconds (defaults to 0.5)'
    parser.add_argument('--segment', metavar='SECONDS', help=segment_help)

    # Extra validation of arguments
    args = parser.parse_args()
    if args.collect:
        if args.action or args.trigger:
            msg = ('-c/--collect should not be used with -a/--action '
                   'or -t/--trigger')
            raise parser.error(msg)
    if args.segment:
        try:
            segment = float(args.segment)
        except ValueError:
            msg = '--segment must be an integer or float'
            raise parser.error(msg)
        if segment < 0.05:
            msg = '--segment cannot be smaller than 0.05'
            raise parser.error(msg)
    if args.action:
        if not args.trigger:
            msg = 'must specify -t/--trigger when using -a/--action'
            raise parser.error(msg)
        if args.action == 'stop':
            args.script = None
        if len(args.trigger) == 1:
            args.trigger.append('1')
        elif len(args.trigger) > 2:
            trigger_msg = '-t/--trigger accepts at most two arguments'
            raise parser.error(trigger_msg)
        else:
            trigger_msg = ('the second argument NUM to -t/--trigger must be an'
                           ' positive integer')
            if not args.trigger[1].isdigit():
                raise parser.error(trigger_msg)
            if args.trigger[1].isdigit() and int(args.trigger[1]) == 0:
                raise parser.error(trigger_msg)

    elif args.trigger:
        msg = 'must specify -a/--action when using -t/--trigger'
        raise parser.error(msg)

    elif args.script:
        if '-e' in sys.argv or '--exec' in sys.argv:
            msg = ('must specify -a/--action and -t/--trigger when using '
                   '-e/--exec')
            raise parser.error(msg)
        else:
            args.script = None
    return args


def get_meter_kwargs():
    kwargs = dict(parse_args()._get_kwargs())
    # Convert `trigger' into `threshold' and `num'
    if kwargs['trigger'] is not None:
        kwargs['threshold'] = kwargs['trigger'][0]
        kwargs['num'] = int(kwargs['trigger'][1])
    del kwargs['trigger']
    kwargs['script'] = get_file_path(kwargs['script'])
    kwargs['log'] = get_file_path(kwargs['log'])
    kwargs['segment'] = float(kwargs['segment']) \
        if kwargs['segment'] is not None else None
    return kwargs


def setup_user_dir():
    if not os.path.exists(USER_DIR):
        os.makedirs(USER_DIR)
    if not os.path.exists(USER_SCRIPT):
        content = '!/bin/sh\n'
        create_executable(USER_SCRIPT, content)
